src/crush/CrushLocation.cc

```diff
@@ -43,14 +45,120 @@ int CrushLocation::_parse(const std::string& s)
+      }
+
+      seastar::experimental::spawn_parameters params = {
+        .argv = {
+          local_conf().get_val<std::string>("crush_location_hook"),
+          "--cluster",
+          local_conf()->cluster,
+          "--id",
+          local_conf()->name.get_id(),
+          "--type",
+          local_conf()->name.get_type_str()
+        }
+      };
+      return seastar::experimental::spawn_process(local_conf().get_val<std::string>("crush_location_hook"), params).then([this] (auto process) {
+        auto stdout = process.stdout();
+        return do_with(std::move(process), std::move(stdout), [this](auto& process, auto& stdout) {
+          return stdout.read().then([] (seastar::temporary_buffer<char> buf) {
+            auto out = std::string(buf.get(), buf.size());
+            boost::algorithm::trim_if(out, boost::algorithm::is_any_of(" \n\r\t"));
+            return seastar::make_ready_future<std::string>(std::move(out));
+          }).then([&process, this] (auto out) {
+            return process.wait().then([out = std::move(out), this] (auto wstatus) {
+              auto* exit_signal = std::get_if<seastar::experimental::process::wait_signaled>(&wstatus);
+              if (exit_signal != nullptr) {
+                std::stringstream errmsg;
+                errmsg << "the user define crush location hook: " << local_conf().get_val<std::string>("crush_location_hook")
+                       << " terminated, terminated signal is " << exit_signal->terminating_signal;
+                logger().error("{}", errmsg.str());
+                throw std::runtime_error(errmsg.str());
+              }
+
+              auto* exit_status = std::get_if<seastar::experimental::process::wait_exited>(&wstatus);
+              if (exit_status->exit_code != 0) {
+                std::stringstream errmsg;
+                errmsg << "the user define crush location hook: " << local_conf().get_val<std::string>("crush_location_hook")
+                       << " execute failed, exit_code is " << exit_status->exit_code;
+                logger().error("{}", errmsg.str());
+                throw std::runtime_error(errmsg.str());
+              } else {
+                _parse(out);
+              }
+              return seastar::now();
+            });
+          });
+        });
+      });
+    });
+  });
+}
+
+seastar::future<> CrushLocation::init_on_startup()
+{
+  if (local_conf().get_val<std::string>("crush_location").length()) {
+    update_from_conf();
+    return seastar::now();
+  }
+  if (local_conf().get_val<std::string>("crush_location_hook").length()) {
+    return update_from_hook();
+  }
+
+  // start with a sane default
+  char hostname[HOST_NAME_MAX + 1];
+  int r = gethostname(hostname, sizeof(hostname));
+  if (r < 0)
+    strcpy(hostname, "unknown_host");
+  // use short hostname
+  for (unsigned i=0; hostname[i]; ++i) {
+    if (hostname[i] == '.') {
+      hostname[i] = '\0';
+      break;
+    }
+  }
+  std::lock_guard l(lock);
```

Comment:
Crimson uses a dummy_mutex. We can skip the lock_guard.

Reply:
done